File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/orderer/sector-populator.ts
import { BluredImageObject } from '../color-compatibility/blured-image-object.model';
import { ImageObject } from '../color-compatibility/image-object.inteface';
import { EventBus } from '../events/event-bus';
import { Tile } from '../models/tile';
import { TileTransform } from '../models/tile-transform.model';
import { IntersectionRecord } from '../sectors/intersections/intersection-record';
import { SectionModel } from '../sectors/section.model';
import { SectorModel } from '../sectors/sector.model';
import { SectorTileTray } from '../tray/section-tile-tray';
import { PossibleLayout } from './section-layout/possible-layout';
import { SectionPopulator } from './section-populator';

export class SectorPopulator {
    private readonly pixelSourceForSector: BluredImageObject;

    public constructor(
        imageObject: ImageObject,
        sector: SectorModel,
        private sectorFilteredTileSource: SectorTileTray,
        private eventBus: EventBus
    ) {
        this.sectorFilteredTileSource = sectorFilteredTileSource;
        this.pixelSourceForSector = this.getPixelSourceForSector(sector, imageObject);
    }

    public populateSection(section: SectionModel): TileTransform[] {
        const sectionPopulator: SectionPopulator = new SectionPopulator(this.pixelSourceForSector, section);

        const avalibleTiles: Tile[] = this.sectorFilteredTileSource.getAvalibleTilesForSector();
        const bestLayout: PossibleLayout = sectionPopulator.searchForBestSectionLayout(avalibleTiles);

        this.markBestSectionLayout(bestLayout);
        this.eventBus.emitSectionOrderingEnd(section);

        return bestLayout.getTilesLayout();
    }

    private markBestSectionLayout(layout: PossibleLayout): void {
        const generatedTiles: TileTransform[] = layout.getTilesLayout();
        layout.section.directTiles = generatedTiles;

        this.sectorFilteredTileSource.removeTiles(generatedTiles.map((tileTransform) => tileTransform.tile));

        for (const singleTileRate of layout.layoutTilesRates) {
            for (const sectionOccupance of singleTileRate[1].sectionsOcupance) {
                const section: SectionModel = sectionOccupance[0];
                const tileTransform: TileTransform = singleTileRate[0];
                const intersectionArea: number = sectionOccupance[1];
                section.intersections.addIntersection(new IntersectionRecord(tileTransform, intersectionArea));
            }
        }
    }

    private getPixelSourceForSector(sector: SectorModel, imageObject: ImageObject): BluredImageObject {
        const worldRadial: number = Math.max(sector.tileMaxRadius / 2, sector.tileMinRadius) / 2;
        return imageObject.getBluredImage(worldRadial, sector.contour);
    }
}

File: backend/Irrtirer/Irrtirer.Generator/Orderer/SectorPopulator.cs
using Irrtirer.Generator.ColorCompatibility;
using Irrtirer.Generator.Helpers;
using Irrtirer.Generator.Models;
using Irrtirer.Generator.MosaicStructure;
using Irrtirer.Generator.Tiles;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Threading;
using Irrtirer.Generator.Orderer.SectionLayout;
using Microsoft.Extensions.Logging;
using System.Linq;

namespace Irrtirer.Generator.Orderer
{
    internal class SectorPopulator
    {
        private readonly EventBus eventBus;
        private readonly RandomFactory randomFactory;
        private readonly SectorTileTray sectorFilteredTileSource;
        private readonly BluredImageObject pixelSourceForSector;
        private readonly ParallelOptions parallelOptions;
        private readonly ILogger logger;

        public SectorPopulator(
            ILogger logger,
            RandomFactory randomFactory,
            IImageSource imageObject,
            EventBus eventBus,
            CancellationTokenSource cancellationTokenSource,
            SectorModel sector,
            SectorTileTray sectorFilteredTileSource)
        {
            this.logger = logger;
            this.eventBus = eventBus;
            this.randomFactory = randomFactory;
            this.sectorFilteredTileSource = sectorFilteredTileSource;

            pixelSourceForSector = GetPixelSourceForSector(sector, imageObject);
            parallelOptions = new ParallelOptions()
            {
                MaxDegreeOfParallelism = 5,
                CancellationToken = cancellationTokenSource.Token
            };
        }

        public List<TileTransform> PopulateSection(SectionModel section)
        {
            SectionPopulator sectionPopulator = new SectionPopulator(logger, randomFactory, pixelSourceForSector, parallelOptions, section);

            Tile[] avalibleTiles = sectorFilteredTileSource.GetAvalibleTilesForSector();
            PossibleLayout bestLayout = sectionPopulator.SearchForBestSectionLayout(avalibleTiles);

            MarkBestSectionLayout(bestLayout);

            eventBus.OnSectionOrderingEnd(section);
            return bestLayout.TilesLayout;
        }

        private void MarkBestSectionLayout(PossibleLayout layout)
        {
            layout.Section.DirectTiles = layout.TilesLayout;

            sectorFilteredTileSource.RemoveTiles(layout.TilesLayout.Select(tileTransform => tileTransform.Tile));

            foreach (var singleTileRate in layout.LayoutTilesRates)
            {
                foreach (var sectionOccupance in singleTileRate.Value.SectionsOcupance)
                {
                    SectionModel section = sectionOccupance.Key;
                    section.Intersections.AddIntersection(new IntersectionRecord()
                    {
                        Tile = singleTileRate.Key,
                        Area = sectionOccupance.Value
                    });
                }
            }
        }

        private BluredImageObject GetPixelSourceForSector(SectorModel sector, IImageSource imageObject)
        {
            float worldRadial = MathF.Max(sector.TileMaxRadius / 2, sector.TileMinRadius) / 2;
            return imageObject.GetBluredImage(worldRadial, sector.Contour);
        }
    }
}
